networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_routing_driver.py

```diff
@@ -0,0 +1,255 @@
+        net_mask = self._get_interface_subnet_from_hosting_port(port,
+            is_external=is_external)
+        sub_interface = self._get_interface_name_from_hosting_port(port)
+        self._do_create_sub_interface(sub_interface, vlan, vrf_name, hsrp_ip,
+                                      net_mask, is_external)
+        # Always do HSRP
+        if ri.router.get(ha.ENABLED, False):
+            if port.get(ha.HA_INFO) is not None:
+                self._add_ha_hsrp(ri, port, is_external=is_external)
+            else:
+                # We are missing HA data, candidate for retrying
+                params = {'r_id': ri.router_id, 'p_id': port['id'],
+                          'port': port}
+                raise cfg_exc.HAParamsMissingException(**params)
+
+    def _add_ha_hsrp(self, ri, port, is_external=False):
+        priority = None
+        if ri.router.get(ROUTER_ROLE_ATTR) in (ROUTER_ROLE_HA_REDUNDANCY,
+                                               ROUTER_ROLE_GLOBAL):
+            for router in ri.router[ha.DETAILS][ha.REDUNDANCY_ROUTERS]:
+                if ri.router['id'] == router['id']:
+                    priority = router[ha.PRIORITY]
+        else:
+            priority = ri.router[ha.DETAILS][ha.PRIORITY]
+        port_ha_info = port[ha.HA_INFO]
+        group = port_ha_info['group']
+        ip = self._get_interface_ip_from_hosting_port(port,
+            is_external=is_external)
+        if is_external:
+            ha_ip = port_ha_info['ha_port']['fixed_ips'][0]['ip_address']
+        else:
+            # TODO(tbachman): needs fixing
+            ha_ip = netaddr.IPAddress(netaddr.IPAddress(ip).value + 1).format()
+        vlan = self._get_interface_vlan_from_hosting_port(port)
+        if ip and group and priority:
+            vrf_name = self._get_vrf_name(ri)
+            sub_interface = self._get_interface_name_from_hosting_port(port)
+            self._do_set_ha_hsrp(sub_interface, vrf_name,
+                                 priority, group, ha_ip, vlan)
+
+    def _add_tenant_net_route(self, ri, port):
+        if self._fullsync and (ri.router_id in
+                               self._existing_cfg_dict['routes']):
+            LOG.debug("Tenant network route already exists, skipping")
+            return
+        cidr = port['subnets'][0]['cidr']
+        if cidr:
+            vrf_name = self._get_vrf_name(ri)
+            out_itfc = self._get_interface_name_from_hosting_port(port)
+            ip = netaddr.IPNetwork(cidr)
+            subnet, mask = ip.network.format(), ip.netmask.format()
+            next_hop = self._get_interface_next_hop_from_hosting_port(port)
+            conf_str = aci_asr1k_snippets.SET_TENANT_ROUTE_WITH_INTF % (
+                vrf_name, subnet, mask, out_itfc, next_hop)
+            self._edit_running_config(conf_str, 'SET_TENANT_ROUTE_WITH_INTF')
+
+    def _remove_tenant_net_route(self, ri, port):
+        cidr = port['subnets'][0]['cidr']
```

Comment:
no need to check "if self._fullsync and (ri.router_id not in self._existing_cfg_dict['routes'])" similar to _add_tenant_net_route()?